src/utilities/aggregateLogs.js

```diff
@@ -0,0 +1,117 @@
+/**
+ * mSupply Mobile
+ * Sustainable Solutions (NZ) Ltd. 2019
+ */
+
+/**
+ * Get daily intervals between two dates.
+ *
+ * @param  {Date}  startDate
+ * @param  {Date}  endDate
+ */
+const getDailyIntervals = (startDate, endDate) => {
+  const dates = [];
+  const date = startDate;
+  while (date <= endDate) {
+    dates.push(new Date(date.getTime()));
+    date.setDate(date.getDate() + 1);
+  }
+
+  return dates;
+};
+
+/**
+ * Get hourly intervals between two dates.
+ *
+ * @param  {Date}  startDate
+ * @param  {Date}  endDate
+ */
+const getHourlyIntervals = (startDate, endDate) => {
+  const dates = [];
+  const date = startDate;
+  while (date <= endDate) {
+    dates.push(new Date(date));
+    date.setHours(date.getHours() + 1);
+  }
+
+  return dates;
+};
+
+/**
+ * Aggregates sensor logs into groups of logs by dates.
+ *
+ * @param   {Realm.results, Object[]}  data                An array of sensorLog objects to
+ *                                                         aggregate.
+ * @param   {boolean}                  isMax               If true, aggregate by max temperature,
+ *                                                         else min temperature.
+ * @param   {number}                   numberOfDataPoints  Number of aggregated data points to
+ *                                                         return.
+ * @param   {Date}                     startDate           Start date of temperature range to
+ *                                                         aggregate over.
+ * @param   {Date}                     endDate             End date of temperature range to
+ *                                                         aggregate over.
+ * @return  {Object[]}                                     Aggregated sensor logs, array of objects
+ *                                                         in form [{timestamp}, {temperature},...].
+ */
+export const aggregateLogs = ({ data, isMax, numberOfDataPoints, startDate, endDate }) => {
+  // If start date is later than earliest date in data, update.
```

Comment:
Hmm i was thinking from the start to try and use the simplicity of realm.
* Sort sensorLogs with realm
* if startDate is not defined use sortedSensorLogs[first]
* if endData is not defined use sortedSensorLogs[last]
* (endDate - startDate) / numberOfDataPoints -> interval in milliseconds
* from 0 to numberOfDataPoints, make an array of object ( [{ startDate, endDate, middleDate}, { startDate, endDate, middleDate}, ... ]
* use realm, data.filtered, (timestamp > startDate and endDate =< endDate).min('temperature') for every element of that array.

done

at the end call a method that checks the size of the interval and formats each middle date accordingly (i.e. hours or minutes etc..) can leave that to the last

Reply:
No worries, I'll have another go at it the Realm way 😆 .